I2c scan recovery reset fix

gcopeland opened the pull request:
# What does this implement/fix?

This fixes i2c's bus recovery logic, especially when combined with scan, to avoid WDT reboot.

Current i2c bus recovery logic has an arbitrary (based on comment) 500ms delay when attempting to recover the i2c bus when the clock line is held low. This both violates the esphome mandate of no delays longer than 10ms, but also triggers the WDT, ensuring recovery is never possible.

This bug was discovered because of an issue with a wiring harness which resulted in the i2c bus clock line held low. While this does prevent recovery of the bus, it also prevents continued operation of the device without the associated i2c bus as it simply boot loops because of the WDT.

This fix does several things. One, it uses the related cycles count as a basis for retry timing. Two, attempts many more retries for recovery. Three, arbitrary reduces the net retry duration below that of the esphome delay mandate. Four, resets the WDT while within the clock recovery logic. 

Technically the reset is not required (untested) as is, however, the pre-existing comment says the 500ms is arbitrary, is rather large, and I do not know if larger delays are required for universal compatibility with i2c devices which may be slow to respond. Which means either the cycle count and/or the delay can be changed well above the 10ms esphome mandate if required, to universally support all i2c devices which may hold low the bus longer than the current values (I defer to i2c expert here). Regardless, removal of the WDT logic is on the table. As is modification to the recovery delay/cycle count.

## Types of changes

- [ x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

## Test Environment

- [x] ESP32
- [] ESP32 IDF
- [x] ESP8266 *compiled
- [ ] RP2040

Tested against ESP32, arduino.
Compiled against ESP8266, arduino.
Code also modifies IDF variant. Not compiled. Not tested.

## Example entry for `config.yaml`:
i2c:
  # For debugging, set scan to true. For production
  # set scan to 'false'.
  scan: true
  frequency: 10kHz

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

Tested against ESP32, arduino.
Compiled against ESP8266, arduino.
Code also modifies IDF variant. Not compiled. Not tested.

Testing on ESP32, with the clock line pulled low confirms the follow resolution:
1. board no longer reboots because of WDT - allowing progression of scan and recovery attempts
2. bus scan now correctly continues and fails, logging the failures of address detection
3. the i2c bus fails and associated devices on the bus are ignored
4. esphome continues running on the node with the associated failed bus/devices.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

gcopeland committed:
Fix for I2C bus recovery w/scanning enabled

The long delay of 20*25ms (500ms) causes the WDT to reset
when the clock line is held low. This means the recovery
never completes. This change drastically reduces the
delay for bus recovery checks (HIGH) and resets the WDT
while attempting to do so. Total delay is roughly 1ms.
If the bus does not reset in a timely manner, it's
doubtful it will. Regardless, this avoids the WDT reset.

gcopeland committed:
Upped the count to allow for i2c bus recovery.

This now waits roughly 2.5ms for the bus to recover
when the clock line is held down. Larger delays are now
possible because the WDT is reset during the loop.

gcopeland committed:
ESP8266 specific changes for WDT reset

This breaks out the ESP8266 WDT reset logic as it is distinct
from that of the ESP32's implementation.

gcopeland committed:
Fixed tidy reference and spacing demands

Added namespace to esp8266 WDT reset call.

gcopeland committed:
Fixes ESP8266 WDT Reference

Messed up my testing of the compile on the ESP8266. This
fixes the issue and compiles against the 8266-ardunio base.

gcopeland committed:
Deleted a space for the conditional USE_ESP8266

Deleted a single space on the ESP822 conditional compilation.
It's picky.

jesserockz commented on esphome/components/i2c/i2c_bus_arduino.cpp:
```diff
@@ -227,10 +230,21 @@ void ArduinoI2CBus::recover_() {
+    // all.
+    auto wait = 250;
     while (wait-- && digitalRead(scl_pin_) == LOW) {  // NOLINT
+#if defined(USE_ESP32)
+      esp_task_wdt_reset();
+      delayMicroseconds(half_period_usec * 2);
+#elif defined(USE_ESP8266)
+      EspClass::wdtFeed();
+      delayMicroseconds(half_period_usec * 2);
+#elif
       delay(25);
+#endif
```
`App.feed_wdt()` exists as a layer so component code doesnt need to mostly check the chip.

```suggestion
      App.feed_wdt();
      delayMicroseconds(half_period_usec * 2);
```

jesserockz commented on esphome/components/i2c/i2c_bus_arduino.cpp:
```diff
@@ -5,6 +5,9 @@
 #include "esphome/core/helpers.h"
 #include <Arduino.h>
 #include <cstring>
+#if defined(USE_ESP32)
+#include <esp_task_wdt.h>
+#endif
```
```suggestion
#include "esphome/core/application.h";
```

jesserockz commented on esphome/components/i2c/i2c_bus_esp_idf.cpp:
```diff
@@ -273,10 +274,14 @@ void IDFI2CBus::recover_() {
+    auto wait = 250;
     while (wait-- && gpio_get_level(scl_pin) == 0) {
-      delay(25);
+      esp_task_wdt_reset();
```
I know this is idf only, but the function actually does a little bit more.
```suggestion
      App.feed_wdt();
```

jesserockz commented on esphome/components/i2c/i2c_bus_esp_idf.cpp:
```diff
@@ -6,6 +6,7 @@
 #include "esphome/core/helpers.h"
 #include <cstring>
 #include <cinttypes>
+#include <esp_task_wdt.h>
```
```suggestion
#include "esphome/core/application.h";
```

gcopeland committed:
Simplification without conditional compilation. Good stuff.

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

gcopeland committed:
Changed to include application layer rather than low level esp headers. Required for App reference.

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

gcopeland committed:
More code simplifcation

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

gcopeland committed:
More code simplification

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

gcopeland committed:
Merge branch 'dev' into i2c_scan_recovery_reset_fix

gcopeland committed:
Removed ; from suggested changes and header order

Removed ';' and changed header include order to more closely
follow the established include ordering.
